src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs

```diff
@@ -332,71 +272,35 @@ private void DoSimultaneousWrites(Func<string, Task> write)
             countdown.Wait();
         }
 
-        [Fact]
-        public async Task PersistentService_Solution_SimultaneousReads()
+        [Theory]
+        [CombinatorialData]
+        public async Task PersistentService_Solution_SimultaneousReads(Size size, bool withChecksum)
         {
             var solution = CreateOrOpenSolution();
-            await PersistentService_Solution_SimultaneousReads(solution, Size.Small);
-            await PersistentService_Solution_SimultaneousReads(solution, Size.Medium);
-            await PersistentService_Solution_SimultaneousReads(solution, Size.Large);
-        }
-
-        private async Task PersistentService_Solution_SimultaneousReads(Solution solution, Size size)
-        {
-            await PersistentService_Solution_SimultaneousReads(solution, size, withChecksum: false);
-            await PersistentService_Solution_SimultaneousReads(solution, size, withChecksum: true);
-        }
-
-        private async Task PersistentService_Solution_SimultaneousReads(Solution solution, Size size, bool withChecksum)
-        {
             var streamName1 = "PersistentService_Solution_SimultaneousReads1";
 
             using var storage = GetStorage(solution);
             Assert.True(await storage.WriteStreamAsync(streamName1, EncodeString(GetData1(size)), GetChecksum1(withChecksum)));
             DoSimultaneousReads(async () => ReadStringToEnd(await storage.ReadStreamAsync(streamName1, GetChecksum1(withChecksum))), GetData1(size));
         }
 
-        [Fact]
-        public async Task PersistentService_Project_SimultaneousReads()
+        [Theory]
+        [CombinatorialData]
```

Comment:
>CombinatorialData [](start = 9, length = 17)

What is this? I can't find a good reference for it.

Reply:
It's a helper from @AArnott. It generates combinations of inputs (e.g. a `bool` parameter would run with both `true` and `false`, and an enum parameter would run with each enum value).

https://github.com/AArnott/Xunit.Combinatorial